pkg/attributes/attributes.go

```diff
@@ -0,0 +1,353 @@
+package attributes
+
+import (
+	"encoding/json"
+	"errors"
+
+	apiext "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+)
+
+// Attributes provides a way to add a map of arbitrary YAML/JSON
+// objects.
+type Attributes map[string]apiext.JSON
+
+// GetDecodedInto allows decoding the attribute with the given key
+// into a given interface. The provided interface should be a pointer
+// to a struct, to an array, or to any simple type.
+//
+// An error is returned if the provided interface type is not compatible
+// with the attribute content
+func (attributes Attributes) GetDecodedInto(key string, into interface{}) error {
+	var err error
+	if attribute, exists := attributes[key]; exists {
+		err = json.Unmarshal(attribute.Raw, into)
+	} else {
+		err = errors.New("Key '" + key + "' doesn't exist")
+	}
+	return err
+}
+
+// Get allows returning the attribute with the given key
+// as an interface. The underlying type of the returned interface
+// depends on the JSON/YAML content of the attribute. It can be either a simple type
+// like a string, a float64 or a bool, either a structured type like
+// a map of interfaces or an array of interfaces.
+//
+// An optional error holder can be passed as an argument
+// to receive any error that might have occured during the attribute
+// decoding
+func (attributes Attributes) Get(key string, errorHolder ...*error) interface{} {
+	if attribute, exists := attributes[key]; exists {
+		container := &[]interface{}{}
+		err := json.Unmarshal([]byte("[ "+string(attribute.Raw)+" ]"), container)
+		if err != nil && len(errorHolder) > 0 && errorHolder != nil {
```

Comment:
I think we should return the error instead of passing in a struct optionally. This approach makes it very easy to hide ignoring a potential error.

Reply:
The whole point of this approach is to avoid requiring a multiple value return, in order to be able to chain calls when we know there is no reason for an error.

Reply:
But here I can change it. I'd be more reluctant to change it for the `GetString()`... methods, since it's quite useful to be able to chain them or directly use the result in an expression.

Reply:
> when we know there is no reason for an error

Can we identify when we know when there is a reason for the error and when there is no?
Like in the shared example https://github.com/amisevsk/devworkspace-operator/compare/use-v1alpha2...davidfestal:use-v1alpha2-and-new-attributes#diff-cee3fd6d4114264b1f235c57bd27c71a9c40cd809efa214a395f5e4cfee8a92fR80
is it safe to propagate an empty string in case the user misconfigured protocol or path as an object?
It shows that such an aproach forces client not to check the erorr. As an alternative, we could have two separate methods(for each type) one would return the error, another will ignore the error but would be named in the way client understand error is suppressed. Like GetStringOrDefault(), default depends on type: string - "", int - 0, interface - {}.

Reply:
More relevant case, where some boolean value is mistyped by user, like public: "true", instead of public: true. Then user will get unexpected behavior instead of error:
```go
func TestBooleanAsString(t *testing.T) {
	stringMap := Attributes{}.FromStringMap(map[string]string{
		"public": "true",
	})
	publicAttr := stringMap.GetBoolean("public")
	assert.Equal(t, true, publicAttr)
}
```